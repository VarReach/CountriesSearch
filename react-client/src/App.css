html {
  box-sizing: border-box;
}

body {
  margin: 0;
}

#root {
  background-color: #221f1f;
}

* {
  box-sizing: inherit;
  font-family: 'Open Sans', sans-serif;
  color: #f9f9f9;
}

.hide {
  display: none !important;
}