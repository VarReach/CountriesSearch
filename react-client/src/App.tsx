import React, { FC, useState, useEffect } from "react";
import CountriesService, { Country, SearchType } from "./Services/CountriesService";
import QS from "query-string";
// Components
import MainPage from "./Components/MainPage";
import SearchPage from "./Components/SearchPage";

type SearchContextValue = {
  fetchCountries: (type: SearchType, value: string) => void;
}
const SearchContext = React.createContext<SearchContextValue>({ fetchCountries: () => null });

export const App: FC = () => {
  const [countries, setCountries] = useState<Country[]>();
  const [error, setError] = useState<Error>();
  const [loading, setLoading] = useState<boolean>(false);

  const fetchCountries = async (type: SearchType, value: string) => {
    if (!loading) setLoading(true);
    try {
      const countries = await CountriesService.get(type, value);
      setCountries(countries);
    } catch (err) {
      setError(err);
    } finally {
      setLoading(false);
    }
  }

  const updateURL = (queries: { type?: SearchType, value?: string}) => {
    const queryString = QS.stringify(queries);
    const newUrl = new URL(window.location.href);
    newUrl.search = `?${queryString}`;
    window.history.pushState(null, "", newUrl.href);
  }

  useEffect(() => {
    // Check for valid Querie params -> execute the search
    if (window.location.search) {
      // Ignore unrelated query params
      const queries = QS.parse(window.location.search);
      const { type } = queries;
      let { value } = queries;
      // Check to make sure Type is valid, if it's not, clear it
      if (type !== "code" && type !== "name" && type !== "fullName") {
        updateURL({ ...queries, type: undefined });
      } else if (type && value) {
        // Use last value if multiple are provided
        value = Array.isArray(value) ? value[value.length - 1] : value;
        // Remove additional values from URL
        updateURL({ ...queries, value });
        fetchCountries(type, value);
      }
    }
  }, []);

  return (
    <SearchContext.Provider value={{ fetchCountries }}>
      {loading || (countries && countries.length) ? 
      <SearchPage /> : 
      <MainPage />
      }
    </SearchContext.Provider>
  );
}

export default App;