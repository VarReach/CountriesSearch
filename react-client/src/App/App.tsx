import React, { FC } from "react";

export const App: FC = () => {
  return (<div>Testing</div>);
}

export default App;