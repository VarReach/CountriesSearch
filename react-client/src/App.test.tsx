import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import CountryService from "./Services/CountriesService";
import QS from "query-string";
import mockCountries from "../__mocks__/countriesMock.js";
// Components
import App from "./App";

/**
 * Covers tests for the App
 */

describe("<App/>", () => {
  let params;

  beforeEach(() => {
    params = {};
    CountryService.get = jest.fn().mockResolvedValue(mockCountries);
    QS.parse = jest.fn().mockReturnValue(params);
  });

  describe("no query params", () => {
    test("renders main page without crashing", () => {
      const { container } = render(<App />);
      expect(screen.getByTestId("main")).toBeTruthy();
      expect(container).toMatchSnapshot();
    });
  });

  describe("query params", () => {
    describe("invalid query params should not be respected", () => {
      test("both invalid", () => {
        params = { type: "invalid" };
        QS.parse = jest.fn().mockReturnValue(params);
  
        render(<App />);
        expect(CountryService.get).not.toHaveBeenCalled();
        expect(screen.getByTestId("main")).toBeTruthy();
      });

      test("type invalid", () => {
        params = { type: "invalid", value: "params"};
        QS.parse = jest.fn().mockReturnValue(params);
  
        render(<App />);
        expect(CountryService.get).not.toHaveBeenCalled();
        expect(screen.getByTestId("main")).toBeTruthy();
      });

      test("valid type = 'code': value length is invalid", () => {
        params = { type: "code", value: "params"};
        QS.parse = jest.fn().mockReturnValue(params);
  
        render(<App />);
        expect(CountryService.get).not.toHaveBeenCalled();
        expect(screen.getByTestId("main")).toBeTruthy();
      });
    });

    describe("valid query params should be respected", () => {
      test("valid: type = 'name'", async () => {
        params = { type: "name", value: "name-value"};
        QS.parse = jest.fn().mockReturnValue(params);
  
        render(<App />);
        expect(CountryService.get).toHaveBeenCalledWith(params.type, params.value);
        const resultsPage = await waitFor(() => screen.getByTestId("results"));
        expect(resultsPage).toBeTruthy();
      });

      test("valid: type = 'fullName'", async () => {
        params = { type: "fullName", value: "name-value"};
        QS.parse = jest.fn().mockReturnValue(params);
  
        render(<App />);
        expect(CountryService.get).toHaveBeenCalledWith(params.type, params.value);
        const resultsPage = await waitFor(() => screen.getByTestId("results"));
        expect(resultsPage).toBeTruthy();
      });

      test("valid: type = 'code'", async () => {
        params = { type: "code", value: "yes"};
        QS.parse = jest.fn().mockReturnValue(params);
  
        render(<App />);
        expect(CountryService.get).toHaveBeenCalledWith(params.type, params.value);
        const resultsPage = await waitFor(() => screen.getByTestId("results"));
        expect(resultsPage).toBeTruthy();
      });
    })
  });
});