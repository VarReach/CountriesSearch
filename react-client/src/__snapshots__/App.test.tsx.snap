// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> no query params renders main page without crashing 1`] = `
<div>
  <div
    class="main"
    data-testid="main"
  >
    <div
      class="main__search"
    >
      <img
        class="logo"
        src="test-file-stub"
      />
      <form
        class="search_form search_form__main"
        id="search_form__main"
      >
        <input
          autocapitalize="off"
          autocomplete="off"
          autocorrect="off"
          class="search_form__input"
          id="search_form__input_main"
          tabindex="1"
          value=""
        />
        <button
          class="search_form__clear search_form__button hide"
          id="search_form__clear"
          type="button"
        >
          <i
            class="material-icons"
          >
            clear
          </i>
        </button>
        <div
          class="search_form__inputs"
        >
          <button
            class="search_form__type-button search_form__button"
            id="search_form__type-button_main"
            type="button"
          >
            <span>
              name
            </span>
            <i
              class="material-icons"
            >
              expand_more
            </i>
          </button>
          <button
            class="search_form__submit search_form__button"
            id="search_form__submit_main"
          >
            <i
              class="material-icons"
            >
              search
            </i>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
